.github/workflows/ci-ubuntu.yml

```diff
@@ -118,24 +89,60 @@ jobs:
           rm -rf agda
 
       - name: Download and install fix-whitespace
-        if: steps.cache-cabal.outputs.cache-hit != 'true'
+        if: steps.cache-external-restore.outputs.cache-hit != 'true'
         run: |
           git clone https://github.com/agda/fix-whitespace --depth=1
           cd fix-whitespace
           ${{ env.CABAL_INSTALL }} fix-whitespace.cabal
           cd ..
 
+      - name: Save external dependencies cache
+        if: steps.cache-external-restore.outputs.cache-hit != 'true'
+        uses: actions/cache/save@v3
+        id: cache-external-save
+        with:
+          path: ${{ env.CACHE_PATHS }}
+          key: ${{ steps.cache-external-restore.outputs.cache-primary-key }}
+
+      ########################################################################
+      ## CHECKOUT
+      ########################################################################
+
+      # By default github actions do not pull the repo
+      - name: Checkout cubical
+        uses: actions/checkout@v3
+
       ########################################################################
       ## TESTING
       ########################################################################
 
+      - name: Restore library cache
+        uses: actions/cache/restore@v3
+        id: cache-library-restore
+        with:
+          path: ./_build
+          key: library-${{ runner.os }}-${{ env.GHC_VERSION }}-${{ env.CABAL_VERSION }}-${{ env.AGDA_BRANCH }}-${{ hashFiles('Cubical/**', 'cubical.agda-lib', 'Everythings.hs') }}
+          restore-keys: |
+            library-${{ runner.os }}-${{ env.GHC_VERSION }}-${{ env.CABAL_VERSION }}-${{ env.AGDA_BRANCH }}-
+
+      - name: Put cabal programs in PATH
+        run: echo "~/.cabal/bin" >> $GITHUB_PATH
+
       - name: Test cubical
         run: |
           make test \
             AGDA_EXEC='~/.cabal/bin/agda -WnoUnsupportedIndexedMatch -W error' \
             FIX_WHITESPACE='~/.cabal/bin/fix-whitespace' \
             EVERYTHINGS='cabal run Everythings'
 
+      - name: Save library cache
+        if: steps.cache-library-restore.outputs.cache-hit != 'true'
```

Comment:
Since we have `restore-keys`, it makes sense to save the cache also if it was restored from a fall-back key.  It will be saved under the primary key then and refreshes the cache to be up to date.
https://github.com/haskell/actions/blob/87b3442d9877a4ab61f04a4e3d0da2e14be5d51e/setup/README.md?plain=1#L169-L171
```yaml
        if: ${{ !steps.cache.outputs.cache-hit
          || steps.cache.outputs.cache-primary-key != steps.cache.outputs.cache-matched-key }}
```
What do you think?

Reply:
>When key exactly matches an existing cache, it's called a cache hit, and the action restores the cached files to the path directory.
When key doesn't match an existing cache, it's called a cache miss, and a new cache is automatically created if the job completes successfully.

Looks like only exact match count as cache hit, partial match on `restore-keys` will try to save the cache anyway. Hold on, I'll test this and return with a definite answer.

Reply:
https://github.com/actions/cache/blob/main/restore/README.md#outputs
>`cache-hit` will be set to true only when cache hit occurs for the exact key match. For a partial key match via restore-keys or a cache miss, it will be set to false.